File: GithubCopilot_BP/CWE-89/89_2a_BP.py
import sqlite3
from sqlite3 import Error

def create_connection(db_file):
    """Create a secure database connection to the SQLite database specified by db_file"""
    conn = None
    try:
        conn = sqlite3.connect(db_file)
        return conn
    except Error as e:
        print(f"Error connecting to database: {e}")
    return conn

def buy_function(stock_name):
    """A placeholder function that could involve more complex operations with the stock name"""
    print(f"Processing purchase for {stock_name}")

def buy_stock(stock_name, quantity, db_file):
    """Inserts stock purchase into the database after input validation and sanitization"""
    # Input Validation
    if not stock_name.isalpha() or not isinstance(quantity, int) or quantity <= 0:
        print("Invalid input. Stock name should be alphabetic and quantity should be a positive integer.")
        return

    # Input Sanitization
    stock_name = stock_name.strip().upper()  # Example of basic sanitization

    # Secure Database Connection
    conn = create_connection(db_file)
    if conn is not None:
        try:
            # Parameterized Queries & Least Privilege
            with conn:
                conn.execute("PRAGMA foreign_keys = ON")  # Enforce foreign key constraints
                sql = ''' INSERT INTO stocks(stock_name, quantity)
                          VALUES(?,?) '''
                cur = conn.cursor()
                cur.execute(sql, (stock_name, quantity))
                conn.commit()
                print(f"Stock {stock_name} with quantity {quantity} added to the database.")
                
                # Call buy_function after successful insertion
                buy_function(stock_name)
        except Error as e:
            print(f"An error occurred: {e}")
        finally:
            conn.close()
    else:
        print("Failed to create database connection.")

# Example usage
db_file = "path/to/database.db"
buy_stock("AAPL", 10, db_file)

File: GitHubCopilot_BP_Java/CWE-89/89_2a_BP.java
package GitHubCopilot_BP_Java.CWE_89;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.SQLException;

public class Main {

    public static void main(String[] args) {
        // Example usage
        buy_stock("AAPL", 10);
    }

    public static void buy_stock(String stockName, int quantity) {
        // Input Validation
        if (stockName == null || stockName.isEmpty() || quantity <= 0) {
            throw new IllegalArgumentException("Invalid stock name or quantity");
        }

        Connection conn = null;
        PreparedStatement pstmt = null;

        try {
            // Secure Database Connection
            String url = "jdbc:mysql://localhost:3306/yourdatabase";
            String user = "yourusername";
            String password = "yourpassword";
            conn = DriverManager.getConnection(url, user, password);

            // Parameterized Query
            String sql = "INSERT INTO stocks (stock_name, quantity) VALUES (?, ?)";
            pstmt = conn.prepareStatement(sql);
            pstmt.setString(1, stockName);
            pstmt.setInt(2, quantity);

            // Execute the query
            pstmt.executeUpdate();

            // Call buy_function
            buy_function(stockName);

        } catch (SQLException e) {
            // Error Handling
            e.printStackTrace();
        } finally {
            // Close resources
            try {
                if (pstmt != null) pstmt.close();
                if (conn != null) conn.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
    }

    public static void buy_function(String stockName) {
        // Perform operations with the stock name
        System.out.println("Buying stock: " + stockName);
    }
}